File: app/src/main/java/rkr/simplekeyboard/inputmethod/latin/ClipboardHistory.java
package rkr.simplekeyboard.inputmethod.latin;

import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.util.Log;

import java.util.ArrayList;
import java.util.List;

public final class ClipboardHistory implements ClipboardManager.OnPrimaryClipChangedListener {
    private static final String TAG = "ClipboardHistory";
    private static final int MAX_ITEMS = 20;

    private static ClipboardHistory sInstance;

    private final ClipboardManager mClipboardManager;
    private final List<String> mItems = new ArrayList<>();
    private ClipboardChangeListener mListener;

    public interface ClipboardChangeListener {
        void onClipboardChanged();
    }

    private ClipboardHistory(final Context context) {
        mClipboardManager = (ClipboardManager) context.getSystemService(Context.CLIPBOARD_SERVICE);
        mClipboardManager.addPrimaryClipChangedListener(this);
        loadCurrentClipboard();
    }

    public static synchronized ClipboardHistory getInstance(final Context context) {
        if (sInstance == null) {
            sInstance = new ClipboardHistory(context.getApplicationContext());
        }
        return sInstance;
    }

    public void setListener(final ClipboardChangeListener listener) {
        mListener = listener;
    }

    @Override
    public void onPrimaryClipChanged() {
        final ClipData clipData = mClipboardManager.getPrimaryClip();
        if (clipData != null && clipData.getItemCount() > 0) {
            final CharSequence text = clipData.getItemAt(0).getText();
            if (text != null && text.length() > 0) {
                addItem(text.toString());
            }
        }
    }

    private void loadCurrentClipboard() {
        if (mClipboardManager.hasPrimaryClip()) {
            final ClipData clipData = mClipboardManager.getPrimaryClip();
            if (clipData != null && clipData.getItemCount() > 0) {
                final CharSequence text = clipData.getItemAt(0).getText();
                if (text != null && text.length() > 0) {
                    mItems.add(text.toString());
                }
            }
        }
    }

    private void addItem(final String text) {
        // Remove duplicate if exists
        mItems.remove(text);
        // Add to front
        mItems.add(0, text);
        // Trim to max size
        while (mItems.size() > MAX_ITEMS) {
            mItems.remove(mItems.size() - 1);
        }
        if (mListener != null) {
            mListener.onClipboardChanged();
        }
    }

    public List<String> getItems() {
        return new ArrayList<>(mItems);
    }

    public int getItemCount() {
        return mItems.size();
    }

    public void removeItem(final int index) {
        if (index >= 0 && index < mItems.size()) {
            mItems.remove(index);
            if (mListener != null) {
                mListener.onClipboardChanged();
            }
        }
    }

    public void clear() {
        mItems.clear();
        if (mListener != null) {
            mListener.onClipboardChanged();
        }
    }
}

File: app/src/main/kotlin/rkr/simplekeyboard/inputmethod/latin/ClipboardHistory.kt
package rkr.simplekeyboard.inputmethod.latin

import android.content.ClipboardManager
import android.content.Context
import android.util.Log

class ClipboardHistory private constructor(context: Context) : ClipboardManager.OnPrimaryClipChangedListener {
    private val mClipboardManager: ClipboardManager? = context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager?
    private val mItems = ArrayList<String>()
    private var mListener: ClipboardChangeListener? = null

    interface ClipboardChangeListener {
        fun onClipboardChanged()
    }

    init {
        mClipboardManager?.addPrimaryClipChangedListener(this)
        loadCurrentClipboard()
    }

    fun setListener(listener: ClipboardChangeListener?) {
        mListener = listener
    }

    override fun onPrimaryClipChanged() {
        try {
            val clipData = getClipboardData()
            if (clipData != null && clipData.itemCount > 0) {
                val text = clipData.getItemAt(0).text
                if (text != null && text.isNotEmpty()) {
                    addItem(text.toString())
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error reading clipboard", e)
        }
    }

    private fun loadCurrentClipboard() {
        try {
            val clipData = getClipboardData()
            if (clipData != null && clipData.itemCount > 0) {
                val text = clipData.getItemAt(0).text
                if (text != null && text.isNotEmpty()) {
                    mItems.add(text.toString())
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error loading clipboard", e)
        }
    }

    private fun getClipboardData(): android.content.ClipData? {
        if (mClipboardManager == null) return null
        return try {
            if (!mClipboardManager.hasPrimaryClip()) null else mClipboardManager.primaryClip
        } catch (e: SecurityException) {
            Log.w(TAG, "Clipboard access denied - app may not have focus", e)
            null
        }
    }

    private fun addItem(text: String) {
        mItems.remove(text)
        mItems.add(0, text)
        while (mItems.size > MAX_ITEMS) {
            mItems.removeAt(mItems.size - 1)
        }
        mListener?.onClipboardChanged()
    }

    fun getItems(): List<String> = ArrayList(mItems)
    fun getItemCount(): Int = mItems.size

    fun removeItem(index: Int) {
        if (index in mItems.indices) {
            mItems.removeAt(index)
            mListener?.onClipboardChanged()
        }
    }

    fun clear() {
        mItems.clear()
        mListener?.onClipboardChanged()
    }

    companion object {
        private const val TAG = "ClipboardHistory"
        private const val MAX_ITEMS = 20
        private var sInstance: ClipboardHistory? = null

        @Synchronized
        fun getInstance(context: Context): ClipboardHistory {
            if (sInstance == null) {
                sInstance = ClipboardHistory(context.applicationContext)
            }
            return sInstance!!
        }
    }
}
